### ISSUE TITLE
Able to capture notes or comments while acquiring locks on a particular file in a repository and retrieve same with Fetch Locks REST API

### ISSUE THREAD

**Issue by pratap043** [git-lfs/git-lfs#5068]
**Describe the issue**
We have a requirement where we need to display the Locks owned by a user in a particular workspace. 
In our development scenarios we have multiple workspaces where same repo will be cloned and worked on further. So during team development we acquire lock on a file before modifying it, so that other users will not be allowed to modify the same file from their workspaces. 

We acquire the lock with the Create Lock REST API [**https://<git-server.com>/foo/bar.git/info/lfs/locks**] by passing the file path and the Fully-qualified server refspec.

**Create Lock Request Json :**
{
    "path": "Project2/Test.java",
    "ref": {
        "name": "refs/heads/master"
    }
}

**Response Json:**
{
    "lock": {
        "id": "1553703",
        "path": "Project2/Test.java",
        "locked_at": "2022-06-23T08:33:40Z",
        "owner": {
            "name": "Abc Xyz"
        }
    }
}

Here we get the information of the lock which is successfully created. In this scenario we don't know from which workspace the lock was created in my requirement scenario. 

So our requirement is to, send some additional info having the workspace info in either notes or comments custom attribute to the Lock REST API and get it back.

Say Workspace A has locks on File A, now when I run GET Fetch Locks REST API [**https://<git-server.com>/foo/bar.git/info/lfs/locks**] from Workspace B, then with the current API, we would get the below response which doesn't have information about the workspace.

**Fetch Locks API Json Response:**
{
    "locks": [
        {
            "id": "1553703",
            "path": "Project2/Project2#cws-prj#.cws",
            "owner": {
                "name": "Abc Xyz"
            },
            "locked_at": "2022-06-23T08:33:40Z"
        },
    ],
    "next_cursor": ""
}

**Expected:**
So it would be helpful if we have some custom attribute while creating locks to store some additional information and then retrieve the same when we access Locks REST API as given below:

**Create Lock Request Json :**
{
  "path": "Project2/Test.java",
  "ref": {
    "name": "refs/heads/master"
  },
  **"notes" : "Lock applied from Workspace A"**
}

**Create Lock Response Json:**
{
    "lock": {
        "id": "1553703",
        "path": "Project2/Test.java",
        "locked_at": "2022-06-23T08:33:40Z",
        "owner": {
            "name": "Abc Xyz"
        },
   **"notes": "Lock applied from Workspace A"**
    }
}

**Fetch Locks Response Json :**
{
    "locks": [
        {
            "id": "1553703",
            "path": "Project2/Project2#cws-prj#.cws",
            "owner": {
                "name": "Abc Xyz"
            },
            "locked_at": "2022-06-23T08:33:40Z",
    **"notes": "Lock applied from Workspace A"**
        },
    ],
    "next_cursor": ""
}

With this enhancement, we can have some predefined comment as part of lock creation and get back same with the Fetch Locks REST API so that it will be useful to differentiate and get more information on the File lock.

**System environment**
http.sslcainfo=C:/Git2.36/mingw64/ssl/certs/ca-bundle.crt
http.sslbackend=openssl
diff.astextplain.textconv=astextplain
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.required=true
filter.lfs.process=git-lfs filter-process
core.autocrlf=true
core.fscache=true
core.symlinks=false
core.editor="C:\\Program Files\\Notepad++\\notepad++.exe" -multiInst -notabbar -nosession -noPlugin
pull.rebase=false
init.defaultbranch=master
user.name=<username>
user.password=<PAT>
user.username=<username>
user.email=<user@abc.com>
safe.directory=C:/AWP/defaultInst/cws/sync/system/GITWorkspace001
safe.directory=C:/AWP/defaultInst/cws/sync/system/GITWorkspace002

**Output of `git lfs env`**
git-lfs/3.1.4 (GitHub; windows amd64; go 1.17.8)
git version 2.36.1.windows.1

LocalWorkingDir=
LocalGitDir=
LocalGitStorageDir=
LocalMediaDir=lfs\objects
LocalReferenceDirs=
TempDir=lfs\tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
GIT_EXEC_PATH=C:/Git2.36/mingw64/libexec/git-core
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"

**Additional context**

**Comment by pratap043** [comment]
Hi Brian, thanks for your quick response.

In our current system, we can have the below use cases, where there are 3 workspaces with 2 users working on a single Repository containing 6 files.

Here the User1 is working on FileA, FileB under Workspace W1, User2 is working on FileC, FileD under Workspace W2 and also User1 is working on FileE, FileF under Workspace W3. 

So we acquire locks before working on these files respectively.  Now we can see that W1 has locks on FileA, FileB. W2 has locks on FileC, FileD. W3 has locks on FileE, FileF.

Repository
	Wl
		User1
			FileA
			FileB
	W2
		User2
			FileC
			FileD
	W3
		User1
			FileE
			FileF
			
			
Now when we run fetchLocks request from W1, we will get all locks on the Repo as given below. Here we doesn't know from which workspaces, the locks were acquired.

User     Workspace    File
----------------------------
User1   	 	        FileA
User1   		        FileB
User1			FileE
User1			FileF
User2			FileC
User2			FileD

Our requirement is to get the workspace information along with locks, to know from where the locks were acquired as given below (say When fetchLocks is run from W1):

User   Workspace    File
----------------------------
User1       W1	 	FileA
User1       W1		FileB
User1	W3		FileE
User1	W3		FileF
User2	W2		FileC
User2	W2		FileD

So if we capture this workspace info as part of custom attribute during lock creation, same can be retrieved back and displayed.

**Comment by pratap043** [comment]
Hi @KalleOlaviNiemitalo, that would be great. Yes, our application would run the lock command by providing the information in the additional attribute which is message in this case. We are using the REST API to acquire the Locks, so we would provide that additional attribute information (message) as part of the request and initiate the Lock creation request programmatically and would like to get back same information with Fetch Locks REST API.

**Comment by pratap043** [comment]
Hi @KalleOlaviNiemitalo, our sample pre-defined comment during acquiring lock is as given below:

&lt;SCMComment xmlns='http://schemas.cordys.com/cws/1.0'&gt; &lt;Comment&gt;Automatically locked by the web interface upon editing&lt;/Comment&gt;&lt;UserDN&gt;sysadmin&lt;/UserDN&gt;&lt;Workspace&gt;'Workspace A' from organization 'system'&lt;/Workspace&gt;&lt;DateTime&gt;1656321671318&lt;/DateTime&gt;&lt;/SCMComment&gt;

Here **UserDN** has the Logged in username of the Application. The owner attribute from the [List Locks ](https://github.com/git-lfs/git-lfs/blob/v3.2.0/docs/api/locking.md#list-locks) response would give us the Git Repo user.  The **Workspace** would give us the workspace information.

We would always need this **SCMComment** info and yes this information can be sent only when the user asks for it.
These comments are generated by our Application software and hence there would be no manual interventions to modify the comments.

So, based on the above comment info, do you see any possibility for enhancing the GIT LFS REST API.

If this comment still leads to a Risk, then we will come up with simpler comment structure.

**Comment by bk2204** [comment]
I agree this sounds like it would be achievable.  Can you open a PR with documentation to `docs/proposals` that would describe the extension to the locking API?  Once that's open, we can ping `@git-lfs/implementers` to see if they'd be willing to implement this feature.

**Comment by pratap043** [comment]
Thanks @bk2204 and @KalleOlaviNiemitalo. I have created a PR with documentation to docs/proposals #[5079](https://github.com/git-lfs/git-lfs/pull/5079)
